program-analysis/echidna/exercises/exercise4/solution.sol

```diff
@@ -0,0 +1,35 @@
+ contract Ownership{
+    address owner = msg.sender;
+    function Owner() public {
+         owner = msg.sender;
+     }
+     modifier isOwner(){
+         require(owner == msg.sender);
+         _;
+      }
+   }
+  contract Pausable is Ownership{
+     bool is_paused;
+     modifier ifNotPaused(){
+              require(!is_paused);
+              _;
+      }
+      function paused() isOwner public{
+          is_paused = true;
+      }
+      function resume() isOwner public{
+          is_paused = false;
+      }
+   }
+   contract Token is Pausable{
+      mapping(address => uint) public balances;
+      function transfer(address to, uint value) ifNotPaused public{
+
+           assert(balances[msg.sender] - value <= balances[msg.sender]);
```

Comment:
The properties in the exercises are:
- `Add an assert to check that, after calling transfer the msg.sender address cannot have more tokens than its initial balance.`
- `Add an assert to check that, after calling transfer the to address cannot have less tokens than its initial balance.`

Should it be more something like
```solidity
uint initial_balance_from = balances[msg.sender];
uint initial_balance_to = balances[to];
// ....
assert(balances[msg.sender] <= initial_balance_from);
assert(balances[to] >= initial_balance_to);
```
We could also simplify the properties to something that will match a simple implementation:
- After calling transfer the msg.sender address must have its initial balance or less
- After calling transfer the to address must have its initial balance or more

Reply:
You are right. These are simpler and easier to understand. Please take a look again.